File: src/features/volume.py
"""Volume and participation features."""

import logging
from typing import Any

import pandas as pd

from .base import EPS, BaseFeatureCalculator, FeatureSpec, safe_divide, zscore

logger = logging.getLogger(__name__)


class VolumeFeatureCalculator(BaseFeatureCalculator):
    """Computes volume and participation features.

    Features:
        - vol1: Raw volume
        - dvol1: Dollar volume (close * volume)
        - relvol_60: Relative volume vs rolling mean
        - vol_z_60: Volume z-score
        - dvol_z_60: Dollar volume z-score

    Participation features help confirm momentum continuation.
    """

    def __init__(self, window: int = 60):
        """Initialize VolumeFeatureCalculator.

        Args:
            window: Rolling window size for z-scores and relative volume
        """
        self.window = window

    @property
    def feature_specs(self) -> list[FeatureSpec]:
        return [
            FeatureSpec(
                name="vol1",
                description="Raw volume",
                lookback=1,
                group="volume",
            ),
            FeatureSpec(
                name="dvol1",
                description="Dollar volume: close * volume",
                lookback=1,
                group="volume",
            ),
            FeatureSpec(
                name="relvol_60",
                description="Relative volume: V / mean(V, 60)",
                lookback=self.window,
                group="volume",
            ),
            FeatureSpec(
                name="vol_z_60",
                description="Volume z-score over 60-bar window",
                lookback=self.window,
                group="volume",
            ),
            FeatureSpec(
                name="dvol_z_60",
                description="Dollar volume z-score over 60-bar window",
                lookback=self.window,
                group="volume",
            ),
        ]

    def compute(self, df: pd.DataFrame, **kwargs: Any) -> pd.DataFrame:
        """Compute volume features.

        Args:
            df: DataFrame with columns: close, volume

        Returns:
            DataFrame with columns: vol1, dvol1, relvol_60, vol_z_60, dvol_z_60
        """
        logger.debug("Computing volume features for %d rows", len(df))
        result = pd.DataFrame(index=df.index)

        # Raw volume
        result["vol1"] = df["volume"]

        # Dollar volume
        result["dvol1"] = df["close"] * df["volume"]

        # Relative volume (current vs rolling mean)
        vol_mean = df["volume"].rolling(window=self.window, min_periods=self.window).mean()
        result["relvol_60"] = safe_divide(df["volume"], vol_mean)

        # Volume z-score
        result["vol_z_60"] = zscore(df["volume"], window=self.window)

        # Dollar volume z-score
        result["dvol_z_60"] = zscore(result["dvol1"], window=self.window)

        logger.debug(
            "Volume features computed: relvol_60 range=[%.4f, %.4f], vol_z_60 range=[%.4f, %.4f]",
            result["relvol_60"].min(), result["relvol_60"].max(),
            result["vol_z_60"].min(), result["vol_z_60"].max()
        )
        return result

File: indicator-engine/src/indicators/volume.cpp
#include "indicators/volume.h"
#include <spdlog/spdlog.h>

namespace ie {

void VolumeCalculator::compute(
    const std::vector<OHLCVBar>& bars,
    std::vector<IndicatorResult>& results) const
{
    const int n = static_cast<int>(bars.size());

    RollingStats vol_stats, dvol_stats;
    vol_stats.init(window);
    dvol_stats.init(window);

    RollingSum vol_sum;
    vol_sum.init(window);

    for (int i = 0; i < n; i++) {
        auto& r = results[i].features;
        double v = static_cast<double>(bars[i].volume);
        double c = bars[i].close;
        double dv = c * v;  // dollar volume

        r["vol1"] = v;
        r["dvol1"] = dv;

        // relvol_60: volume / mean(volume, 60)
        vol_sum.push(v);
        if (vol_sum.full()) {
            r["relvol_60"] = safe_divide(v, vol_sum.mean());
        } else {
            r["relvol_60"] = NaN;
        }

        // vol_z_60: zscore of volume
        vol_stats.push(v);
        r["vol_z_60"] = vol_stats.full() ? vol_stats.zscore(v) : NaN;

        // dvol_z_60: zscore of dollar volume
        dvol_stats.push(dv);
        r["dvol_z_60"] = dvol_stats.full() ? dvol_stats.zscore(dv) : NaN;
    }

    spdlog::debug("VolumeCalculator: computed {} bars", n);
}

} // namespace ie
